### PR TITLE
Add gp200r support 

### PR DESCRIPTION
**Pull request by drolleigh**
Created a support package for gp200r including a urdf that has been compared to gp200r pendant read-outs of a real robot in unique poses. The translation was accurate under 1.0 mm and rotation was accurate under 1.0 degree.

gp200r_pendant_urdf_compare

### CODE REVIEW

File: motoman_gp200r_support/package.xml
Comment (gavanderhoorn) [inline motoman_gp200r_support/package.xml]: is this really where you got the information?

If you used a document, please cite the document title and document number instead.

File: motoman_gp200r_support/package.xml
Comment (gavanderhoorn) [inline motoman_gp200r_support/package.xml]: Is this copy-pasta from the GP180 package?

File: motoman_gp200r_support/urdf/gp200r_macro.xacro
Comment (gavanderhoorn) [inline motoman_gp200r_support/urdf/gp200r_macro.xacro]: Could you please use the `material` defs from `motoman_resources` instead?

File: motoman_gp200r_support/urdf/gp200r_macro.xacro
Comment (gavanderhoorn) [inline motoman_gp200r_support/urdf/gp200r_macro.xacro]: Please do not use `scale`.

Instead, scale the meshes to use metres.

File: motoman_gp200r_support/urdf/gp200r_macro.xacro
Comment (gavanderhoorn) [inline motoman_gp200r_support/urdf/gp200r_macro.xacro]: Could you clarify where you got the `effort` values from?

File: motoman_gp200r_support/urdf/gp200r_macro.xacro
Comment (gavanderhoorn) [inline motoman_gp200r_support/urdf/gp200r_macro.xacro]: Please be consistent and use `${radians(..)}` everywhere.

File: motoman_gp200r_support/urdf/gp200r_macro.xacro
Comment (gavanderhoorn) [inline motoman_gp200r_support/urdf/gp200r_macro.xacro]: This appears to use `visual` quality meshes for `collision` purposes. That's not recommended. Performance will suffer, especially in motion planners like MoveIt.

Please create suitable collision meshes. You could create convex hulls using Blender or some other mesh editing tool.

File: motoman_gp200r_support/urdf/gp200r_macro.xacro
Comment (gavanderhoorn) [reply motoman_gp200r_support/urdf/gp200r_macro.xacro]: Please also don't use uppercase characters in filenames (including extensions).

File: motoman_gp200r_support/urdf/gp200r_macro.xacro
Comment (gavanderhoorn) [inline motoman_gp200r_support/urdf/gp200r_macro.xacro]: While I believe I understand why you add the 90 degree rotation over Y here, I'm not sure it's needed.

Did you consider keeping REP-103-aligned orientation and use the X+ axis for the translations of `joint_3_u` and `joint_4_r`?

File: motoman_gp200r_support/urdf/gp200r_macro.xacro
Comment (gavanderhoorn) [inline motoman_gp200r_support/urdf/gp200r_macro.xacro]: Please move `tool0`'s `link` to the end of the file together with its `joint` declaration block (place it after the `base` `link` and `joint`).

Refer to [sia30d_macro.xacro](https://github.com/ros-industrial/motoman/blob/21c83deee272215fc307874ec32447fcffabdbee/motoman_sia30d_support/urdf/sia30d_macro.xacro#L165-L179) for an example.

**Review by gavanderhoorn — CHANGES_REQUESTED** [review]
Thanks for the PR :+1: 

I've left a few in-line comments.

Could you please also take a look at the size of the individual meshes (ie: `.stl`s)? The `visual` directory alone is 6.5MB, with files such as `link_3_u.STL` by itself already approx 1.8MB.

We should try to optimise mesh size, as it prevents repository bloat.

I find Blender with the Decimate modifier a good way to reduce mesh size, while keeping mesh quality.

File: motoman_gp200r_support/urdf/gp200r_macro.xacro
Comment (EricMarcil) [reply motoman_gp200r_support/urdf/gp200r_macro.xacro]: I usually use the Max Acc/Dec Torque as the effort values that I extract from the robot parameters:
Please refer to this file: [Robot_Specs_GP200R-A00.csv](https://github.com/ros-industrial/motoman/files/7990153/Robot_Specs_GP200R-A00.csv)

File: motoman_gp200r_support/urdf/gp200r_macro.xacro
Comment (EricMarcil) [inline motoman_gp200r_support/urdf/gp200r_macro.xacro:146]: We might also want to add a flange frame here, to work towards https://github.com/ros-industrial/motoman/issues/374. Refer to [motoman_hc10_support/urdf/hc10dt_b10_macro.xacro](https://github.com/ros-industrial/motoman/blob/21c83deee272215fc307874ec32447fcffabdbee/motoman_hc10_support/urdf/hc10dt_b10_macro.xacro#L145-L151) for an example.

**Comment by akashjinandra** [comment]
Thanks for all the feedback we will start on these changes and have this updated very soon!

File: motoman_gp200r_support/package.xml
Comment (drolleigh) [reply motoman_gp200r_support/package.xml]: I used decimate in Blender to reduce the .stl files to half the size, and I scaled them before exporting them to be used as visual meshes.

No. I have updated the reference document link for joint limits and maximum joint velocities.

File: motoman_gp200r_support/package.xml
Comment (drolleigh) [reply motoman_gp200r_support/package.xml]: I have updated it to refer to myself and my manager as maintainers.

File: motoman_gp200r_support/urdf/gp200r_macro.xacro
Comment (drolleigh) [reply motoman_gp200r_support/urdf/gp200r_macro.xacro]: I have added a line at the top of the urdf to load the xacro for common materials provided by motoman_resources so that ```"material_yaskawa_blue"``` is used.
```
<xacro:include filename="$(find motoman_resources)/urdf/common_materials.xacro"/>
```

File: motoman_gp200r_support/urdf/gp200r_macro.xacro
Comment (drolleigh) [reply motoman_gp200r_support/urdf/gp200r_macro.xacro]: I used the effort values in the [Robot_Specs_GP200R-A00.csv](https://github.com/ros-industrial/motoman/files/7990153/Robot_Specs_GP200R-A00.csv) doc that Eric provided. Thank you!

File: motoman_gp200r_support/urdf/gp200r_macro.xacro
Comment (drolleigh) [reply motoman_gp200r_support/urdf/gp200r_macro.xacro]: I have changed all angle parameters to match the ```${radians(degrees)}``` form.

File: motoman_gp200r_support/urdf/gp200r_macro.xacro
Comment (drolleigh) [reply motoman_gp200r_support/urdf/gp200r_macro.xacro]: I have exported convex hulled .stl's from Blender to use for collision in the urdf.

File: motoman_gp200r_support/urdf/gp200r_macro.xacro
Comment (drolleigh) [reply motoman_gp200r_support/urdf/gp200r_macro.xacro]: If I understand you and REP-103-aligned orientation correctly, then I have made the necessary changes to ```joint_2_l```, ```joint_3_u``` and ```joint_4_r```. Please let me know if further changes need to be made to the coordinate systems.

File: motoman_gp200r_support/urdf/gp200r_macro.xacro
Comment (drolleigh) [reply motoman_gp200r_support/urdf/gp200r_macro.xacro]: I moved the tool0 link.

File: motoman_gp200r_support/urdf/gp200r_macro.xacro
Comment (drolleigh) [reply motoman_gp200r_support/urdf/gp200r_macro.xacro:146]: I have added a flange frame as done in the referenced hc10 xacro.

File: motoman_gp200r_support/package.xml
Comment (gavanderhoorn) [inline motoman_gp200r_support/package.xml:36]: let's always keep a ROS-I maintainer in the list.

```suggestion
    <maintainer email="g.a.vanderhoorn@tudelft.nl">G.A. vd. Hoorn (TU Delft Robotics Institute)</maintainer>
    <license>BSD</license>
```

File: motoman_gp200r_support/urdf/gp200r_macro.xacro
Comment (gavanderhoorn) [inline motoman_gp200r_support/urdf/gp200r_macro.xacro]: Is this correct? This joint now appears twice.

Comment (drolleigh) [reply motoman_gp200r_support/urdf/gp200r_macro.xacro]: Not supposed to be there twice. I removed the extra joint_6_t-tool0 block, and I pushed to pr which includes the ROS-I maintainer on the list.

Comment (akashjinandra) [reply motoman_gp200r_support/urdf/gp200r_macro.xacro]: Side note @gavanderhoorn having this joint in there twice really messes up moveit collision checking when you add a gripper at the end of tool0. Get errors such as gripper_link_1 is in collision with gripper_link_1. All of those errors go away once we take out that redundant joint.

Comment (akashjinandra) [reply motoman_gp200r_support/urdf/gp200r_macro.xacro]: It's kind of surprising that the URDF allows the user to do that, I thought links were only allowed to have one parent, but technically since the joint is in there twice it has the same parent.....

Comment (gavanderhoorn) [reply motoman_gp200r_support/urdf/gp200r_macro.xacro]: > when you add a gripper at the end of `tool0`

regardless of the problems caused by duplicated `joint`s and `link`s: you're not supposed to attach grippers / EEF models to `tool0`. That's what `flange` is for.

See also [Coordinate Frames: flange](https://gavanderhoorn.github.io/rep/rep-0199.html#flange).

**Comment by gavanderhoorn** [comment]
I've added a few more commits -- mostly cosmetic.

We'll wait for CI to turn green again and then merge.

**Review by gavanderhoorn — APPROVED** [review]
Thanks for the PR @drolleigh.

**Comment by gavanderhoorn** [comment]
Note: we'll use a squash-merge here, as we don't need all the fixup commits to end up in the repository's history.

**Comment by gavanderhoorn** [comment]
Thanks again @drolleigh and @akashjinandra.

**Comment by drolleigh** [comment]
Our pleasure. Glad we could contribute!

**Comment by gavanderhoorn** [comment]
@drolleigh: because I squash-merged, GH may not notify you your branch can be deleted.

**Pull request merged**
